### DISCUSSION TITLE
RFC - Custom Code

### DISCUSSION THREAD

**Discussion by yuval-hazaz** [amplication/amplication#1224] [General] [answered]
# RFC - Custom Code

## Summary
Currently, Amplication provides developers with a fully functional node.JS application with all the following features/ modules:

- Nest.JS as a framework for the server
- Prisma as the ORM with auto-generated schema based on the entities defined in Amplication UI.
- Models (classes) - based on the entities, and fields as defined in Amplication UI
- DTOs (classes) - for all CRUD operations on the models
- Services - Nest.JS services, one for each model (entity), with methods for all CRUD operations on the entity.
- GraphQL API resolvers - for all entities, with CRUD operations
- REST API Controllers - for all entities, with CRUD operations
- Security Layer - with RBAC attributes per action (in resolver & controllers)  and granular permissions for specific fields
- React client (Admin UI) for CRUD operations on all entities

In most or all real-life use cases, the developers are required to add custom business logic on top of the generated code.

This RFC provides a proposal and details on a method to allow developers to alter and customize the generated application without losing the ability to continue using Amplication when changes to are required. 

The method described in this RFC is based on the assumption that the app in Amplication is linked to a GitHub repository which enables code sync from Amplication server to GitHub and from GitHub to Amplication server.

The main obstacle when customizing an application generated with Amplication is to allow the Amplication server to keep adding models, roles, and other generated code to the linked repository without overriding the custom code that was added by the developer. 

## Proposal
The main tool to achieve the required functionality is class inheritance. 
Amplication will generate all services, controllers, and resolvers in two separate classes - a base class, and the main class that extended the base class.

### The method

- All the generated code will be included in the base class,
- The main class will be an empty class that extends the base class.
- Developers will be able to freely alter and customize the main class with any custom code.
- Amplication server will never override the main classes. 
- The developer will be able to add any additional files and classes that will be import or referenced from the main classes. 
- Amplicaiton server will pull the main classes and any additional files from GitHub, in order to be able to deploy the full application to the sandbox environment.
- The base classes will be placed in a separate folder to provide a clear boundary which files can be altered, and which files should not be altered and will be overridden by the Amplication server with every change.

### Current folder structure
Currently, the following is a typical entity folder structure in a generated app

```
src
└── Customer
   ├── CreateCutomerArgs.ts
   ├── DeleteCutomerArgs.ts
   ├── FindManyCutomerArgs.ts
   ├── FindOneCutomerArgs.ts
   ├── customer.controller.spec.ts
   ├── customer.controller.ts
   ├── customer.module.ts
   ├── customer.resolver.ts
   ├── customer.service.ts
   ├── customer.ts
   ├── CustomerCreateInput.ts
   ├── CustomerUpdateInput.ts
   ├── CustomerWhereInput.ts
   ├── CustomerWhereUniqueInput.ts
   └── updateCustomerArgs.ts
     
```

### Proposed folder structure
Following is the new proposed folder structure

```
src
└── Customer
   ├── Base                     # files in this folder will be overridden with new generated code for every change
   │    ├── CreateCutomerArgs.ts
   │    ├── DeleteCutomerArgs.ts
   │    ├── FindManyCutomerArgs.ts
   │    ├── FindOneCutomerArgs.ts
   │    ├── customer.controller.base.spec.ts
   │    ├── customer.controller.base.ts
   │    ├── customer.resolver.base.ts
   │    ├── customer.service.base.ts
   │    ├── customer.ts
   │    ├── CustomerCreateInput.ts
   │    ├── CustomerUpdateInput.ts
   │    ├── CustomerWhereInput.ts
   │    ├── CustomerWhereUniqueInput.ts
   │    └── updateCustomerArgs.ts
   ├── customer.controller.spec.ts      
   ├── customer.controller.ts    # Exteneds ./base/project.controller.base.ts
   ├── customer.module.ts                 
   ├── customer.resolver.ts      # Exteneds ./base/project.resolver.base.ts 
   └── customer.service.ts       # Exteneds ./base/project.service.base.ts 
```  

### Example - customer.service.ts
This is an example of a service that inherits from the base service. The service is generated by Amplication only once, and the developer can freely add methods, functions, imports, dependencies, and anything else that is needed.

``` typeScript
import { Injectable } from "@nestjs/common";
import { CustomerServiceBase } from "./base/customer.service";
import { PrismaService } from "nestjs-prisma";

@Injectable()
export class CustomerService extends CustomerServiceBase {
  constructor(prisma: PrismaService) {
    super(prisma);
  }

  
}

```

## Motivation/Use Cases
The base classes can be altered in any of the following ways:
- Add new functions and methods
   - New API endpoint in  controllers
   - New queries and mutations in resolvers
   - New functions in services
- Override system (generated) functionality  
   - Override API endpoints to execute custom logic instead of the native behavior
   - Override resolvers queries and mutations to execute custom logic instead of the native behavior 
   - Override service functions to execute custom logic, while keeping the existing controllers and resolvers that trigger the functions.
- Block system (generated) functionality  - when needed, the developer may override generated functionality and throw "NotImplemented" exception instead of executing the native behavior.

## Benefits
Developers will be able to freely customize their apps without worrying that their code will be overridden,. 
Developers will have a clear boundary of which files they can alter, and which should not be altered. 

## Issues Addressed
The following issues would be satisfied by implementing this RFC:

## Open Items

1. How to allow developers to add additional npm packages and keep the package.json in sync. (can we use Yarn workspaces for some kind of inheritance?) 
2. How can we allow developers to remove completely REST endpoints, and GraphQL queries/ mutations from the schema? The current suggestion of "NotImplemented Exception" is not good enough as the schema will still reflect the original state.

**Comment by vimota** [comment]
Exciting! This approach seems like a solid way forward.

I would also suggest stating all other major alternatives considered to give a concrete comparison between these and other possible solutions. I include one below:

### Hooks-based
One solution I imagined for this problem is a hooks-based mechanism inspired by React. In one instance of this scenario, Amplication could provide a `/generated` sub-directory where the current auto generated files live and another one initiated as an empty folder where the user places hooks to be picked up / "registered" (or files placed at the root, or a single file that "registers" all hooks like a routes file). 

Benefits:
- Fewer files to clutter the workspace. The only business logic-related files required would be the ones strictly added by the users as opposed to having empty skeleton files.
- A more functional approach which may be more familiar to the modern front-end users. The functional approach may also prove to be better for composability of business logic.
- Avoiding the class-based override system may reduce bugs caused by statefulness of classes.
- **Less tightly couple to how resolvers/services/etc are implemented. Having a functional hooks-based api would allow the internal generated code to evolve from it's current classes without breaking compatibility with user-code.**

Drawbacks:
- The functional approach may not actually gel well with NestJS and the rest of the Amplication framework - mixing two paradigms may cause confusion.
- It may be less obvious to users where they can override and write custom business logic (as opposed to having clearly defined skeleton files to be filled in).
- Hook-based approach could lead to messier code, having "subscriptions"/hooks defined in multiple places (although there's nothing exactly restricting this from happening on the class-based approach).



### Conclusion
I wanted to flesh out this other approach as an exercise in imagination of potential futures for Amplication. I still personally lean towards your proposed approach, but I think the hooks-based model raises some interesting questions to consider for this design - in particular that of Amplication's ability to evolve the generated code without breaking user-code. Would be great to explore what that part of the design will look like.

**Reply by yuval-hazaz** [reply]
I like the functional approach of your proposal. 
Like you mentioned it may cause complications working together with NestJS.
One more thing to consider - we want to keep all the generated code as standard as it can be so developers don't have to learn new skills that are unique to Amplicaiton. Adding a learning curve to the customization process may become a big blocker. 

Looking ahead, we may consider additional methods to allow developers to choose the best way that fits their needs - and the webhook approach is a good candidate.

**Comment by yuval-hazaz** [comment]
two more things that need to be considered:
1. How to allow developers to add additional npm packages and keep the package.json in sync. (can we use Yarn workspaces for some kind of inheritance?) 
2. How can we allow developers to remove completely REST endpoints, and GraphQL queries/ mutations from the schema? The current suggestion of "NotImplemented Exception" is not good enough as the schema will still reflect the original state.

(These questions were also added to the body of the RFC to be addressed )

**Comment by yuval-hazaz** [comment, marked as answer]
Released as version 0.4.0
https://github.com/amplication/amplication/releases/tag/v0.4.0
